server/src/main/java/io/spine/server/route/MessageRouting.java

```diff
@@ -158,13 +192,49 @@ public void remove(Class<? extends M> messageClass) {
     public R apply(M message, C context) {
         checkNotNull(message);
         checkNotNull(context);
-        K messageClass = toMessageClass(message);
-        Route<M, C, R> func = routes.get(messageClass);
-        if (func != null) {
+        @SuppressWarnings("unchecked") Class<? extends M>
+        cls = (Class<? extends M>) message.getClass();
+        Match match = routeFor(cls);
+        if (match.found()) {
+            Route<M, C, R> func = match.route();
             R result = func.apply(message, context);
             return result;
         }
-        R result = getDefault().apply(message, context);
+        R result = defaultRoute().apply(message, context);
         return result;
     }
+
+    /**
+     * Provides information on routing availability.
+     */
+    final class Match {
+
+        private final Class<? extends M> requestedClass;
+        private final @Nullable Route<M, C, R> route;
+        private final @Nullable Class<? extends M> entryClass;
+
+        private Match(Class<? extends M> requestedClass,
```

Comment:
Can you please describe the parameters? It's not 100% clear what a `requestedClass` is.